### PR TITLE
Handle import key errors

### PR DESCRIPTION
**Pull request by Thykof**

### CODE REVIEW

**Review by maxime-carabina — APPROVED** [review]
LGTM

File: pkg/app/app.go [hunk 1]
```diff
@@ -126,3 +126,11 @@ type ImportFromPKey struct {
 func (a *WalletApp) ImportPrivateKey(pkey string, nickname string, password string) {
 	a.PromptInput <- ImportFromPKey{PrivateKey: pkey, Nickname: nickname, Password: password}
 }
+
+func (a *WalletApp) IsAlreadyExists(nickname string) bool {
```
Comment (mazmassa) [inline pkg/app/app.go:130]:
```suggestion
func (a *WalletApp) IsNicknameUnique(nickname string) bool {
```

File: wails-frontend/src/pages/importPrivateKey.tsx [hunk 2]
```diff
@@ -1,97 +1,131 @@
...
+    privateKey ? 'Define your account name' : 'Enter your Private Key';
+
+  const placeholder = () => (privateKey ? 'Account name' : 'Private key');
+
+  async function validateNickname(e: SyntheticEvent) {
+    const formObject = parseForm(e);
+    const { nickname } = formObject;
+
+    if (!nickname) {
+      setError({ nickname: t('nickname-required') });
```
Comment (mazmassa) [inline wails-frontend/src/pages/importPrivateKey.tsx:47]: Non blocking: If we could, as much as possible, assign the key as `page.sub.field-validation` it will be, IMHO, easy to manage when this became big. So, in this case something like 'import.nickname-required`.

File: wails-frontend/src/pages/importPrivateKey.tsx [hunk 3]
```diff
@@ -1,97 +1,131 @@
...
+      state = { ...state, pkey: privateKey, nickname: input };
+      navigate('/new-password', { state });
+    }
+  }
 
-  const updateInput = (e: React.ChangeEvent<HTMLInputElement>) => {
+  function updateInput(e: React.ChangeEvent<HTMLInputElement>) {
     setInput(e.target.value);
-  };
+  }
```
Comment (mazmassa) [inline wails-frontend/src/pages/importPrivateKey.tsx:84]: Non blocking: Since we are doing nothing here and just wrapping a function maybe it's better to setup there!? 🤔 

```tsx
onChange={(e) => setInput(e.target.value)}
```

File: wails-frontend/src/pages/importPrivateKey.tsx [hunk 4]
```diff
@@ -1,97 +1,131 @@
...
-          <div className="pb-4">
-            <Password
-              id="input"
-              onChange={updateInput}
-              autoComplete="off"
-              name="input"
-              placeholder={placeholder()}
-              value={input}
-            />
+        <form ref={form} onSubmit={handleSubmit}>
```
Comment (mazmassa) [inline wails-frontend/src/pages/importPrivateKey.tsx:89]: I guess we are mixing things here. Controlled and Uncontrolled components. If you want to have all data at once use that onSubmit. But since you are using onChange events I guess you don't need it. 

Am I missing something!? 🤔

File: wails-frontend/src/pages/importPrivateKey.tsx (hunk 3 above)
Comment (Thykof) [reply wails-frontend/src/pages/importPrivateKey.tsx:84]: but I use 2 time this function updateInput. I don't want to duplicate `onChange={(e) => setInput(e.target.value)}`

File: wails-frontend/src/pages/importPrivateKey.tsx [hunk 5]
```diff
@@ -1,97 +1,131 @@
...
           </div>
-          <div className="min-w-fit w-full">
-            <Button onClick={handleApply}>{applyStr}</Button>
+          {privateKey ? (
+            <div className="pb-4">
+              <Input
+                onChange={updateInput}
+                autoComplete="off"
+                name="nickname"
+                placeholder={placeholder()}
```
Comment (mazmassa) [inline wails-frontend/src/pages/importPrivateKey.tsx:102]:
```suggestion
                placeholder={getPlaceholder()}
```

File: web-frontend/src/validation/nickname.tsx [hunk 6]
```diff
@@ -8,5 +8,5 @@ export function isAlreadyExists(
 }
 
 export function isNicknameValid(nickname: string): boolean {
-  return /^[a-z0-9_-]+$/.test(nickname);
+  return /^[a-zA-Z0-9_-]+$/.test(nickname);
```
Comment (mazmassa) [inline web-frontend/src/validation/nickname.tsx:11]: What kind of validation this regex is doing!? Could you please add a description above!? You know regex is hard to follow! 😆

File: wails-frontend/src/pages/importPrivateKey.tsx (hunk 2 above)
Comment (mazmassa) [reply wails-frontend/src/pages/importPrivateKey.tsx:47]: Ok! Good for me! Just need to keep track of that. Set some task including the i18N on placeholder. Otherwise we'll miss it! 😅 

Feel free to close here then...

File: wails-frontend/src/pages/importPrivateKey.tsx (hunk 4 above)
Comment (Thykof) [reply wails-frontend/src/pages/importPrivateKey.tsx:89]: right, I added form after

**Review by mazmassa — APPROVED** [review]
Left some comments, but overall LGTM. Once done feel free to close the threads!

**Comment by Thykof** [comment]
@mazmassa I refactored since your last review, could you have a look again?

File: wails-frontend/src/pages/importPrivateKey/PromptPrivateKey.tsx [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+          </div>
+          <div>
+            <p className="mas-body text-neutral pb-4">Enter your Private Key</p>
+          </div>
+          <div className="pb-4">
+            <Password
+              name="privateKey"
+              placeholder={'Private key'}
+              error={error?.privateKey}
+            />
```
Comment (maxime-carabina) [inline wails-frontend/src/pages/importPrivateKey/PromptPrivateKey.tsx:69]:
```suggestion
            <Password
              defaultValue=""
              name="privateKey"
              placeholder={'Private key'}
              error={error?.privateKey}
            />
```

File: wails-frontend/src/pages/importPrivateKey/PromptNickname.tsx [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+            <p className="mas-body text-neutral pb-4">
+              Define your account name
+            </p>
+          </div>
+          <div className="pb-4">
+            <Input
+              name="nickname"
+              placeholder={'Account name'}
+              error={error?.nickname}
+            />
```
[the remark this answers is not in the document]
Comment (maxime-carabina) [reply wails-frontend/src/pages/importPrivateKey/PromptNickname.tsx:83]: You missed to add `defaultValue`. So just add it and you can merge

File: wails-frontend/src/pages/importPrivateKey/PromptPrivateKey.tsx (hunk 7 above)
Comment (maxime-carabina) [reply wails-frontend/src/pages/importPrivateKey/PromptPrivateKey.tsx:69]: Same here

File: wails-frontend/src/pages/importPrivateKey/PromptNickname.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (Thykof) [reply wails-frontend/src/pages/importPrivateKey/PromptNickname.tsx:83]: why do we need default value?

File: web-frontend/src/validation/nickname.tsx [hunk 9]
```diff
@@ -7,6 +7,14 @@ export function isAlreadyExists(
...
+ * This functino validates the nickname using the following rules:
+ * - must contain only alphanumeric characters, underscores and dashes
+ * - must be between 1 and 32 characters long
+ *
+ * @param nickname nickname to validate
+ * @returns true if nickname is valid, false otherwise
+ */
 export function isNicknameValid(nickname: string): boolean {
-  return /^[a-z0-9_-]+$/.test(nickname);
+  return /^[a-zA-Z0-9_-]+$/.test(nickname) && nickname.length <= 32;
```
Comment (mazmassa) [inline web-frontend/src/validation/nickname.tsx:19]:
```suggestion
  // - must have at least 1 character
  // - must contain only alphanumeric characters, underscores and dashes
  // - must not exceed MaxNicknameLength characters
  return /^[a-zA-Z0-9_-]+$/.test(nickname) && nickname.length <= 32;
```

**Review by mazmassa — APPROVED** [review]
Left some comments but overall LGTM!
